File: LeetCode_Algorithm/1091(Shortest Path in Binary Matrix)/1091.py
'''
LeetCode 1091 : Shortest Path in Binary Matrix
n*n 그리드가 존재한다. 최소 길이의 clear path를 구하라.
clear path란 왼쪽 위에서 시작하여 오른쪽 아래로 가는 길을 의미한다.
방문 가능한 노드는 0으로 표시되어 있다.

# sol1 795ms 33% 15.5MB 12%
bfs로 돌면서 확인


'''
class Solution:
    def shortestPathBinaryMatrix(self, grid: list[list[int]]) -> int:
        if grid[0][0] == 1:
            return -1
        directions = [[0,1], [1,1], [1,0], [1,-1], [0, -1], [-1, -1], [-1, 0], [-1, 1]]
        from collections import deque
        queue = deque([(0, 0, 1)]); seen = set((0, 0))
        while queue:
            r, c, cnt = queue.popleft()
            if r == len(grid) - 1 and c == len(grid[0]) - 1:
                return cnt
            for i in range(len(directions)):
                newR = r + directions[i][0]
                newC = c + directions[i][1]
                if 0 <= newR < len(grid) and 0 <= newC < len(grid[0]) and (newR, newC) not in seen:               
                    seen.add((newR, newC))
                    if grid[newR][newC] == 0:
                        queue.append((newR, newC, cnt+1))
        return -1
            

File: LeetCode_Algorithm/1091(Shortest Path in Binary Matrix)/1091.js
var shortestPathBinaryMatrix = function(grid) {
  // sol1 776ms 5% 59.7MB 6%
  if (grid[0][0] == 1)
    return -1
  let directions = [[0,1], [1,1], [1,0], [1,-1], [0,-1], [-1,-1], [-1,0], [-1,1]]
  let queue = [[0, 0, 1]]; let seen = new Set(["0,0"])
  while (queue.length) {
    [r, c, cnt]= queue.shift()
    if (r == grid.length-1 && c == grid[0].length-1)
      return cnt
    for (let [dr, dc] of directions) {
      let newR = r + dr; let newC = c + dc
      if (0 <= newR && newR < grid.length && 0 <= newC && newC < grid[0].length) {
        // console.log(newR, newC, !seen.has(String(newR) + ',' + String(newC) && grid[newR][newC] == 0))
        if (grid[newR][newC] == 0 && !seen.has(String(newR) + ',' + String(newC))) {
          console.log(newR, newC)
          seen.add(String(newR) + ',' + String(newC))
          queue.push([newR, newC, cnt+1])
        }
      }
    }
  }
  return -1
};
console.log(shortestPathBinaryMatrix([[0,0,0],[1,1,0],[1,1,0]]))
// [0,0,0],
// [1,1,0],
// [1,1,0]